Unique Hashes For USDScene Reopening the Same File

danieldresser-ie opened the pull request:
A quick attempt at implementing the approach I discussed with John today.

I initially did actually try to make the hashes more consistent using a map that assigned 0 to the first time a fileName is opened - but then I realized that we use SdfPath's hashes all over the place, so none of the hashes are consistent anyway, so it really makes more sense to just use a global counter.

It seems like in addition to resolving the issue with prototype names swapping around, this also means that it's now OK to use SdfPath::Hash() as we currently are - it seems like SdfPaths are reliably consistent for the lifespan of a USD file. This is based on two observations:
* if this wasn't true, then everything would currently be broken
* a cursory reading of CrateFile::_ReadPaths seems to suggest that the CrateFile does keep all the contained SdfPaths alive

johnhaddon commented on contrib/IECoreUSD/test/IECoreUSD/USDSceneTest.py:
```diff
@@ -474,6 +474,60 @@ def testInstancesShareHashes( self ) :
+
+			self.assertNotEqual( h1, h2 )
+
+			self.assertFalse( h1 in usedHashes )
```
It's a small thing, but `assertNotIn()` gives more informative error messages than `assertFalse()`.

johnhaddon commented on contrib/IECoreUSD/test/IECoreUSD/USDSceneTest.py:
```diff
@@ -474,6 +474,60 @@ def testInstancesShareHashes( self ) :
+				self.assertEqual( h1, instanceJ.child( "world" ).hash( scene.HashType.TransformHash, 1 ) )
+				del instanceJ
+				
+			self.assertFalse( h2 in usedHashes )
```
As above.

danieldresser-ie commented:
Addressed feedback. I also addressed a problem that I had planned to mention yesterday, and forgot: there was one assert in testSetHashes which tested that two different USDScenes opened from the same file had the same hashes. Since it is now our intention that this not be true, I've just removed that failing assert.
